### ISSUE TITLE
PyQt5 QML standalone not working

### ISSUE THREAD

**Issue by brupelo** [kayhayen/Nuitka#76]
Moving an old interesting issue from the nuitka [mailing list](https://www.freelists.org/post/nuitka-dev/PyQt5-QML-standalone-not-working,3) here. If you ask me, posting issues at github feels really good and the UI is really nice and cozy, the other mailing list is just... well, "a very unfriendly and cold way to post issues", sorry :'D

Anyway, here's the steps you need to do if you want to run that qml test using nuitka+virtualenv:

 - git clone https://github.com/Siecje/qml-testing
 - d:\software\python364_32\Scripts\virtualenv.exe --python=d:\software\python364_32\python.exe venv
 - venv\Scripts\activate
 - pip install pyqt5 nuitka
 - set path=D:\sources\personal\python\apps\nuitka_tests\pyqt\qml-testing\venv\Lib\site-packages\PyQt5\Qt\bin;%path%
 - nuitka3 main.py --show-progress --recurse-all --standalone --plugin-enable=qt-plugins
 - cp *.qml *.js main.dist
 - set QML2_IMPORT_PATH=D:\sources\personal\python\apps\nuitka_tests\pyqt\qml-testing\venv\Lib\site-packages\PyQt5\Qt\qml

If you adjust the hardcoded paths (they point out to my local paths) you should be able to get a running main.exe, now, what's the problem here?

The QML dlls are not being bundled with nuitka and as you can see I'm setting the env.var to use my local ones, why is this?

**Comment by kayhayen** [comment]
The difference between the mailing list and the issue tracker is the amount of people who are going to look at it. When you post an issue here, I am going to see it, just as a I see it on the mailing list. But then nobody else is going to look at it with the intent of helping. Depending on the issue, I have had cases where the mailing list was good at providing extra information. Not going to happen on the mailing list.

It varies. But if an issue is not immediately fixed after mailing list, the issue tracker is a logical next step I agree.

This is the code that Nuitka's plugin for Qt, which you are using, uses to extract the path for the Qt plugins:

from __future__ import print_function

```
import PyQt%(qt_version)d.QtCore
for v in PyQt%(qt_version)d.QtCore.QCoreApplication.libraryPaths():
    print(v)
import os
guess_path = os.path.join(os.path.dirname(PyQt%(qt_version)d.__file__), "plugins")
if os.path.exists(guess_path):
    print("GUESS:", guess_path)
```

What does that give you?

**Comment by Siecje** [comment]
The plugin path is not correct. It should be 
```Python
guess_path = os.path.join(os.path.dirname(PyQt%(qt_version)d.__file__), "Qt", "plugins")
```

**Comment by brupelo** [comment]
@kayhayen When I said friendly I meant from an UI perspective point of view, posting nicely formatted code and all these extra features github gives you for free... As for the amount of people who are going to look at it... At the moment the github project gets more traction, you'll see :) , each time I see people asking about python packaging I just recommend them nuitka as the best choice. 

To be honest, I think the fact there are no more contributors yet is:

a) The project requires coders who have certain skills/expertise in a non-trivial field of CS, what does this mean? Probably a lot of casual-users think the project is awesome but they won't be able to contribute to the project because the intrinsic complexity
b) Marketing... The project is awesome, but... why after so many years is not already the most popular tool to package python projects? I see very often people still recommending the other not-so-good choices (pyinstaller and friends...), --> [this](https://i.stack.imgur.com/fvdkb.png) <--

Anyway, the above comments are just pure speculation so they could just completely wrong :) , it's not like I'm a marketing expert in any case :'D

Regarding the issue:

            command = """\
    from __future__ import print_function

    import PyQt%(qt_version)d.QtCore
    for v in PyQt%(qt_version)d.QtCore.QCoreApplication.libraryPaths():
        print(v)
    import os
    guess_path = os.path.join(os.path.dirname(PyQt%(qt_version)d.__file__), "plugins")
    if os.path.exists(guess_path):
        print("GUESS:", guess_path)
    """ % {
               "qt_version" : qt_version
            }

prints:

from __future__ import print_function

    from __future__ import print_function

    import PyQt5.QtCore
    for v in PyQt5.QtCore.QCoreApplication.libraryPaths():
        print(v)
    import os
    guess_path = os.path.join(os.path.dirname(PyQt5.__file__), "plugins")
    if os.path.exists(guess_path):
        print("GUESS:", guess_path)

and the result of `getPyQtPluginDirs` is:

    ['D:\\sources\\personal\\python\\apps\\nuitka_tests\\venv\\lib\\site-packages\\PyQt5\\Qt\\plugins']

**Comment by Siecje** [comment]
Interesting... you actually get the correct path.
I get `'C:\\Users\\Cody\\path\\to\\venv\\lib\\site-packages\\PyQt5\\plugins'`

**Comment by kayhayen** [comment]
In the normal case, the guessed path is ignored when parsing the output:

```
            # Take the guessed path only if necessary.
            if line.startswith("GUESS: "):
                if result:
                    continue

                line = line[len("GUESS: "):]
```

Not sure what this was necessary for. May have been a standalone Python distribution of some sorts. It is from the initial commit of the plugin code, so I better had commented that, but I didn't. Maybe it was the first stabe, then for Python 2.x, and later got replaced by the querying code, which works in more setups. 

There is even the question, why it would ever not work. But it might be a prudent idea to do the guessing for both variants, including the new way.

I even did a harder check for virtualenv, I found one of my Python installs without a system wide PyQt, and did a compilation for PyQt and before executing, I moved the virtualenv away, so it cannot be used, and then it still ran. So this means we don't have any of that issue here. I will look at your reproducer now.

**Comment by kayhayen** [comment]
@brupelo I am also very positive about Github, there actually are more pull requests incoming. My original idea to be on the software that bugs.python.org is using, have failed, largely because many of their customizations to roundup never became available to me.

Now I do have an issue template, and pull request template, and that's probably going to help a lot.

I just merged a WIP for multiprocessing changes of python3.4, something I often would have had to figure out myself.

But I truly think that once the speedups are coming, more people with skills will jump ship. Because for distribution, those other solutions are typically fine.

But this e.g. motivated me to make options for plugins a thing. So now --enable-plugins=qt-plugins=a,b,c is going to work, and will allow to select plugin folders. Getting the plugin system more documented is going to make people work there a lot, and get rid of special software that needs tricks.

I am going to make Github an ever more prominent choice and phase out the bugs.nuitka.net tracker, which while giving me more control, would be too much work to make it usable. It's a pity for me as a free software guy, to use closed source software like this. But then gmail is closed to and serves a similar purpose and is OK by me too.

**Comment by kayhayen** [comment]
Hello there,

turning to this issue now. You wrote:

> The QML dlls are not being bundled with nuitka and as you can see I'm setting the env.var to use my > local ones, why is this?

Is this still true or was it caused by the other issue with things taken from PATH? It might be.

Yours,
Kay

**Comment by brupelo** [comment]
Kay,

Cool, moving on, so... when I'm back home I'll try to create a nice minimal reproducer about this one using your latest factory to see what the status of this one is so it can be fixed ;)

Hold your horses for a little while :)

**Comment by brupelo** [comment]
@kayhayen Hi, today I've had some time to test this factory [revision](https://github.com/kayhayen/Nuitka/commit/d376d37b7657262f4c8dd48eda161ba00e8aa28b) with this example. I've created a test for you to play with, just do:

 - git clone https://github.com/brupelo/nuitka_tests
 - cd nuitka_tests\pyqt\test3
 - venv.bat
 - build.bat
 - cd main.dist && main.exe

At this point you should see the next error `file:///D:/nuitka_tests/pyqt/test3/main.dist/Main.qml:2 plugin cannot be loaded for module "QtQuick.Controls": Cannot load library D:\nuitka_tests\pyqt\test3\venv\Lib\site-packages\PyQt5\Qt\qml\QtQuick\Controls.2\qtquickcontrols2plugin.dll: The specified module could not be found.`. 

Here's a [video](https://dl.dropboxusercontent.com/s/dsso6amgha1kpz1/2018-05-07_15-52-10.mp4) showing the reproducer in action, also you can see as the @Siecje example is working out the box if just running `python main.py` before building the package.

> Is this still true or was it caused by the other issue with things taken from PATH? It might be.

So yeah, it seems the other nice fix you did for #73 won't address this one.

**Comment by brupelo** [comment]
@Siecje I've added set `QML2_IMPORT_PATH` in the scripts so that way it'll skip another different error than the current one. As for `.qml` files... what do you mean? `build.bat` [is already doing so](https://github.com/brupelo/nuitka_tests/blob/master/pyqt/test3/build.bat#L3).

**Comment by Siecje** [comment]
I meant you don't need it for the Python application to work.

Just that Nuikta doesn't automatically know what to do with `.qml` files.

You are getting the same error, just from `Main.qml` trying to load a QML component.

**Comment by brupelo** [comment]
@Siecje In any case, the reported error when running main.exe `file:///D:/nuitka_tests/pyqt/test3/main.dist/Main.qml:2 plugin cannot be loaded for module "QtQuick.Controls": Cannot load library D:\nuitka_tests\pyqt\test3\venv\Lib\site-packages\PyQt5\Qt\qml\QtQuick\Controls.2\qtquickcontrols2plugin.dll: The specified module could not be found` is totally confusing me... because if I do ` ls D:\nuitka_tests\pyqt\test3\venv\Lib\site-packages\PyQt5\Qt\qml\QtQuick\Controls.2\qtquickcontrols2plugin.dll` I can see that dll is living in that directory... so what does it mean `The specified module could not be found` in this context?

That said, the fact couple of weeks ago I'd been able to run main.exe and right now I've got stuck with this error is also puzzling me... I'll just assume couple of weeks ago one of my tests was just wrong and probably i was hacking some env.var like PATH, who knows (let's say I was just lucky :))... Anyway, if you've got some clue about how to run main.exe that can guide Kay in the right direction, please post it, whatever it helps to narrow it down.

**Comment by Siecje** [comment]
Which version of PyQt5? The most recent release doesn't work.

**Comment by brupelo** [comment]
@kayhayen @Siecje Guys, I've adjusted the repo-test so now the nuitka content will work at least locally, it seems I was missing some path tweaks on my .bat files. Here's the [proof](https://dl.dropboxusercontent.com/s/wjjknoulv1g56vp/2018-05-09_00-36-03.mp4) showing that the test should work locally, you can also give it a shot on your side and you should be able to reproduce... (win7 over here)

Anyway, the important question is whether the [generated nuitka main.dist package](https://www.dropbox.com/s/2rcewbi278qiotd/test3.zip?dl=1) will crash or not on your boxes? Could you please confirm that so we know at which point of narrowing this issue down we are atm?

Thx.
B.

**Comment by Siecje** [comment]
The generated Nuitka main.dist package doesn't work because it doesn't include the Qt QML files.

You are setting `QML2_IMPORT_PATH` to a path in the venv, which isn't included in the Nuitka output.
```batch
set QML2_IMPORT_PATH=%~dp0venv\Lib\site-packages\PyQt5\Qt\qml
```

**Comment by brupelo** [comment]
@Siecje Thanks for the report!

@kayhayen Ok, so it seems the [package test](https://www.dropbox.com/s/2rcewbi278qiotd/test3.zip?dl=1) (using hacks) won't work other than locally. Do you have any idea how to make a proper standalone package out of it? I've run out of hacky ideas :)

**Comment by kayhayen** [comment]
A couple of things @brupelo 

In the test3 and build.bat, you are doing this: `set path=%~dp0venv\Lib\site-packages\PyQt5\Qt\bin;%path%` That should do nothing for you, right?

Also, this part is only intended to hide the bug, right? Do not do that. But I think it will have no impact on the commands in  `set QML2_IMPORT_PATH=%~dp0venv\Lib\site-packages\PyQt5\Qt\qml` build.bat, why would it? Any test that would run CPython and Nuitka in a different environment is dangerous in that it can lead to false conclusions.

I think that merely copying the files from that directory "qml" as "main.dist\PyQt5\Qt\qml" will make things work. At which point this becomes an issue of data files that are to be included for "PyQt5.QtQml", for which the plugin would simply have to return their list.

Implementing  `def considerDataFiles(self, module):` for the PyQt/PySide plugin may then be good enough, it is supposed to return a list of "source_path, relative_dist_path" tuples.

It then ought to work. Can you try out if the copying is sufficient, because if it is, I will add the copying of the QtQml things to the plugin.

Also, make sure to use current factory code and "--force" the currently valid optparse shortcut for making sure to ignore the cache one time. The plugins too might be linked against the Python3.DLL I do not know, but I would avoid the topic that way.

Yours,
Kay

**Comment by brupelo** [comment]
@kayhayen Kay, I've seen at the end you've decided to add this debug [option](https://github.com/kayhayen/Nuitka/commit/a9518df7247de169040481f85669e2ce4eb4523f). Regarding to this, I'll be opening another issue to talk you about some improvements to make the existing caching system smarter, to be blunt I must to say right now the cache system is far away of being optimal. I mean, when nuitka hit the cache times are just great but the main problem right now when working with different virtualenvs is there will be quite a lot of miss caches than it should be in reality, anyway, I'll be talking more about it on other issue to improve this fact.

Alright, so I've tried your previous (suggestion](https://github.com/brupelo/nuitka_tests/blob/master/test3/build.bat) but unfortunately I'm still experiencing the error, here's the steps:

 - git clone https://github.com/brupelo/nuitka_tests && cd nuitka_tests
 - venv.bat
 - cd test3
 - build.bat
 - cd main.dist && main.exe

And I'll get this one:

    QQmlApplicationEngine failed to load component
    file:///D:/sources/personal/python/apps/nuitka_tests/test3/main.dist/Main.qml:2 plugin cannot be loaded for module "QtQuick.Controls": Cannot load library D:\sources\personal\python\apps\nuitka_tests\test3\main.dist\PyQt5\Qt\qml\QtQuick\Controls.2\qtquickcontrols2plugin.dll: The specified module could not be found.

Didn't I understand your previous post, I assumed you asked me to test this [build.bat](https://github.com/brupelo/nuitka_tests/blob/master/test3/build.bat) as the old hackish [build_old.bat](https://github.com/brupelo/nuitka_tests/blob/master/test3/build_old.bat) would lead us to false conclusions as you've very well explained before.

Could you please clarify? Did you make it work over there?

Thanks!

**Comment by kayhayen** [comment]
No, I did not make it work. I just shared my observations. There is two approaches, making workarounds, then facing another issue, potentially caused by the workaround, then reaching a point, where people can't go on. I typically get a report of the later. That situation is then barely interesting to me, the first bug actually is.

Did you try copying the "qml" directory as indicated before? I believe it is at least part of your problem, as usual more things might pop up. No, I am not trying it myself once I reached sufficient reason to believe it will improve things, and there is a user to try it, and to get motivated to make the changes necessary to incorporate into Nuitka.

Actually i didn't even attempt this test case myself yet. I saw no reason to, as the missing "qml" directory is pretty obviously part of the problem.

**Comment by brupelo** [comment]
> Did you try copying the "qml" directory as indicated before?

In the previous mail I've attached you a link to this [batch file](https://github.com/brupelo/nuitka_tests/blob/master/test3/build.bat), I'd say so :) . Btw, it's a shame than window mkdir doesn't have -p like posix command (ie: i could have main.dist\PyQt5\Qt\qml) or just avoiding mkdir and copying the whole thing creating dirs automatically, suggestions welcome to improve that batch file :D

> and there is a user to try it, and to get motivated to make the changes necessary to incorporate into Nuitka.

@Siecje Interested on getting more involved on your issue? Tbh, I've move your issue from nuitka mailing list here but I'm not very interested on qml myself, I've used qt(since 2002)/pyqt(the last 5 years) and I've always been a happy man without using qml on any my multiple personal projects :)

@kayhayen Anyway, If you eventually got any other observations to try please let it know and I'll probably give it a shot, thanks.

**Comment by Siecje** [comment]
I'm on Windows 10. I have installed Python 3.6.5 32-bit.

I have installed Visual Studio with the C++ compiler.

I open x86 Native Tools Command Prompt for VS 2017.

I create a venv with 32-bit Python.

```
C:\Users\Cody\Downloads\python-3.6.5-embed-win32\python.exe -m venv venv
call venv\Scripts\activate.bat
python -m pip install pip --upgrade
python -m pip install nuitka
python -m pip install PyQt5
git clone https://github.com/siecje/qml-testing
cd qml-testing
nuitka main.py --recurse-all --plugin-enable=qt-plugins --verbose
```

There is a main.exe in the current directory. There is no `main.dist/`.

In a new Command Prompt if I do `call main.exe` I get

```
C:\Users\Cody\Desktop\threetwo\qml-testing>call main.exe
Traceback (most recent call last):
  File "C:\Users\Cody\Desktop\threetwo\qml-testing\main.py", line 6, in <module>
    from PyQt5 import QtCore, QtGui, QtQml
ImportError: DLL load failed: %1 is not a valid Win32 application.

C:\Users\Cody\Desktop\threetwo\qml-testing>
```

If I do `call main.exe` with the venv being activated then it works.

**Comment by kayhayen** [comment]
@Siecje Do not call Nuitka as "nuitka", but instead with the "python -m nuitka" invocation method. For bug reporting, the effects of PATH could then harm. The reason you are not having a "main.dist" is that do you not specify --standalone, in which case you get an accelerated binary.

The DLL loading is affected by PATH. If you change it, you get another DLL. Accelerated binaries are not nearly intended to work outside the environment they were created. Standalone ones are intended for use on other machines. Changing PATH is a bad idea for that purpose.

The error you get, I have mostly been getting when changing between 32 and 64 bits. And in some cases, called "uninstalled", it will copying the pythonxy.dll to the directory main.exe lives, which in turn can be bad, if you switch Pythons or PATH, or whatever.

Accelerated mode binaries won't suffer from Qt plugin issues. In fact the Nuitka plugin won't do anything in that case as only standalone mode needs help. They are much simpler to handle, as they directly use the Python installation. They however assume for it to not vanish below them, which it did, when you deactivated the venv.

Yours,
Kay

**Comment by kayhayen** [comment]
Looking at this issue right now @brupelo, thanks again for the really good way of reporting. There is two issues:

a) Missing hidden dependency on QtQuick that is in the QML file only. I am adding an implicit dependency by the QML stuff and then I get only this:

b)

> main.dist/Main.qml:2 plugin cannot be loaded for module "QtQuick.Controls": Cannot load library main.dist\PyQt5\Qt\qml\QtQuick\Controls.2\qtquickcontrols2plugin.dll: Module cannot be found

It is there, but likely has not all the required DLLs, as the plugin scan for Qt plugins was limited to DLLs in side of "plugins", but nearby in "qml" there is a lot more to be had:

> main.dist/PyQt5/Qt/qml/Qt/labs/calendar/qtlabscalendarplugin.dll
> main.dist/PyQt5/Qt/qml/Qt/labs/folderlistmodel/qmlfolderlistmodelplugin.dll
> main.dist/PyQt5/Qt/qml/Qt/labs/handlers/handlersplugin.dll
> main.dist/PyQt5/Qt/qml/Qt/labs/platform/qtlabsplatformplugin.dll
> main.dist/PyQt5/Qt/qml/Qt/labs/settings/qmlsettingsplugin.dll
> main.dist/PyQt5/Qt/qml/Qt/labs/sharedimage/sharedimageplugin.dll
> main.dist/PyQt5/Qt/qml/QtBluetooth/declarative_bluetooth.dll
> main.dist/PyQt5/Qt/qml/QtGraphicalEffects/private/qtgraphicaleffectsprivate.dll
> main.dist/PyQt5/Qt/qml/QtGraphicalEffects/qtgraphicaleffectsplugin.dll
> main.dist/PyQt5/Qt/qml/QtLocation/declarative_location.dll
> main.dist/PyQt5/Qt/qml/QtMultimedia/declarative_multimedia.dll
> main.dist/PyQt5/Qt/qml/QtNfc/declarative_nfc.dll
> main.dist/PyQt5/Qt/qml/QtPositioning/declarative_positioning.dll
> main.dist/PyQt5/Qt/qml/QtQml/Models.2/modelsplugin.dll
> main.dist/PyQt5/Qt/qml/QtQml/StateMachine/qtqmlstatemachine.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls/qtquickcontrolsplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls/Styles/Flat/qtquickextrasflatplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls.2/Fusion/qtquickcontrols2fusionstyleplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls.2/Imagine/qtquickcontrols2imaginestyleplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls.2/Material/qtquickcontrols2materialstyleplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls.2/qtquickcontrols2plugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Controls.2/Universal/qtquickcontrols2universalstyleplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Dialogs/dialogplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Dialogs/Private/dialogsprivateplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Extras/qtquickextrasplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Layouts/qquicklayoutsplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/LocalStorage/qmllocalstorageplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Particles.2/particlesplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/PrivateWidgets/widgetsplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Scene2D/qtquickscene2dplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Scene3D/qtquickscene3dplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Shapes/qmlshapesplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Templates.2/qtquicktemplates2plugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/Window.2/windowplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick/XmlListModel/qmlxmllistmodelplugin.dll
> main.dist/PyQt5/Qt/qml/QtQuick.2/qtquick2plugin.dll
> main.dist/PyQt5/Qt/qml/QtSensors/declarative_sensors.dll
> main.dist/PyQt5/Qt/qml/QtTest/qmltestplugin.dll
> main.dist/PyQt5/Qt/qml/QtWebChannel/declarative_webchannel.dll
> main.dist/PyQt5/Qt/qml/QtWebEngine/qtwebengineplugin.dll
> main.dist/PyQt5/Qt/qml/QtWebSockets/declarative_qmlwebsockets.dll

That is a terrible lot of DLLs to also include. I  wonder if it also means that more than QtQuick ought to be a hidden dependency of QtQml module. 

Or if the ".qml" file should enter the main.dist via the Qt-Plugin, and then be scanned for imports, and yield that as dependencies.

For a solution right now, I am adding the inclusion of  these DLLs once "qml" plug-in is copied, I might make "sensible" the default for Qt DLLs, where "qml" might not be considered sensible, so users would have to add it manually to the argument (it would have to be more transparent what it means, and what can be used), and the copying of the QML folder to the proper spot automatically.

That should be only a bit of twiddling after which I have all rights to be optimistic for this to work out of the box, or be able to warn when I see QQmlApplicationEngine being used without including the "qml" plugins.

**Comment by kayhayen** [comment]
So I did some improvements to the plugin copying, and turns out, that works now. I have yet to put this through my internal tests before I will let it land on factory.

Turns out the "qml" plugin lives outside of "plugins" as its own thing. Makes me uncertain if and how this relates to PySide. Maybe the same applies, maybe not. I would like to sort that out, but maybe it is not relevant at all. I will see.
